internal/storage/localdir.go

```diff
@@ -0,0 +1,117 @@
+package storage
+
+import (
+	"archive/tar"
+	"bytes"
+	"compress/gzip"
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+
+	"github.com/nlepage/go-tarfs"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var _ Storage = &LocalDirectory{}
+
+const DefaultBundleCacheDir = "/var/cache/bundles"
+
+var DefaultLocalDirectory = &LocalDirectory{RootDirectory: DefaultBundleCacheDir}
+
+type LocalDirectory struct {
+	RootDirectory string
+}
+
+func (s *LocalDirectory) Load(_ context.Context, owner client.Object) (fs.FS, error) {
+	bundlePath := filepath.Join(s.RootDirectory, fmt.Sprintf("%s.tgz", owner.GetUID()))
+	bundleFile, err := os.Open(bundlePath)
+	if err != nil {
+		return nil, err
+	}
+	defer bundleFile.Close()
+	tarReader, err := gzip.NewReader(bundleFile)
+	if err != nil {
+		return nil, err
+	}
+	return tarfs.New(tarReader)
+}
+
+func (s *LocalDirectory) Store(_ context.Context, owner client.Object, bundle fs.FS) error {
+	buf := &bytes.Buffer{}
+	gzw := gzip.NewWriter(buf)
+	tw := tar.NewWriter(gzw)
+	if err := fs.WalkDir(bundle, ".", func(path string, d fs.DirEntry, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if d.Type()&os.ModeSymlink != 0 {
+			return nil
+		}
+		info, err := d.Info()
+		if err != nil {
+			return fmt.Errorf("get file info for %q: %w", path, err)
+		}
+
+		h, err := tar.FileInfoHeader(info, "")
+		if err != nil {
+			return fmt.Errorf("build tar file info header for %q: %w", path, err)
+		}
+		h.Uid = 0
+		h.Gid = 0
+		h.Uname = ""
+		h.Gname = ""
+		h.Name = path
+
+		if err := tw.WriteHeader(h); err != nil {
+			return fmt.Errorf("write tar header for %q: %w", path, err)
+		}
+		if d.IsDir() {
```

Comment:
I believe that this never happen because the `getObjects` fails if the manifest has a directory.  So it doesn't matter but I'm wondering if the header should be written with out the writing the data for the directory.

Reply:
This is meant to be generic and usable by another provisioner type that uses a different bundle format.

This code is effectively "just tar.gz the bundle and save it." It's up to the provisioner to read the `fs.FS` and decide what's valid or not.

In the case of the plain provisioner, you're right. We use `getObjects` in the bundle controller just to check that what we're about to save is actually a valid plain provisioner. Other provisioners should probably follow that same pattern.